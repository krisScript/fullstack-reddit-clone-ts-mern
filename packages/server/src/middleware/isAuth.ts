import { verify } from 'jsonwebtoken';
import { Request, Response, NextFunction } from 'express';
import { ErrorREST, Errors } from '../utilities/ErrorREST';
const secret: any = process.env.SECRET;
const isAuth = (req: Request, res: Response, next: NextFunction): void => {
  const authHeader = req.get('Authorization');
  const { status, message } = Errors.Unauthorized;
  const error = new ErrorREST(status, message, null);
  console.log(authHeader);
  if (!authHeader) {
    throw error;
  }
  const token = authHeader.split(' ')[1];

  let decodedToken: any;
  try {
    decodedToken = verify(token, secret);
  } catch (err) {
    err.statusCode = 500;
    throw err;
  }
  if (!decodedToken) {
    throw error;
  }
  req.userId = decodedToken.userId;
  next();
};
export default isAuth;
