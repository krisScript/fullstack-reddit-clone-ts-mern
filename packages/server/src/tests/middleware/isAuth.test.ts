import isAuth from '../../middleware/isAuth';
import httpMocks from 'node-mocks-http';
import jwt from 'jsonwebtoken';
describe('isAuth', (): void => {
  it(`should add userId to req`, async (): Promise<void> => {
    const nextMock = jest.fn();
    const userId = 'randomUserId';
    const secret: any = process.env.SECRET;
    const token = jwt.sign(
      {
        userId,
      },
      secret,
      { expiresIn: '1h' },
    );
    const reqMock = httpMocks.createRequest({
      method: 'POST',
      url: '/',
      headers: {
        Authorization: 'Bearer ' + token,
      },
    });
    const resMock = httpMocks.createResponse();
    isAuth(reqMock, resMock, nextMock);
    expect(nextMock).toBeCalledTimes(1);
    expect(reqMock.userId).toEqual(userId);
  });
  it("should throw an error if req doesn't have Authorization header", async (): Promise<
    void
  > => {
    const nextMock = jest.fn();
    const reqMock = httpMocks.createRequest({
      method: 'POST',
      url: '/',
    });
    const resMock = httpMocks.createResponse();
    expect(
      (): void => isAuth(reqMock, resMock, nextMock),
    ).toThrowErrorMatchingSnapshot();
  });
});
