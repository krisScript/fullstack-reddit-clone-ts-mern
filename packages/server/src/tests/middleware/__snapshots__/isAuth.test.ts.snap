// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`isAuth should throw an error if req doesn't have Authorization header 1`] = `"Authentication credentials not valid."`;
